File: jbolo/src/com/jbolo/bolo/GameType.java
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */


/*
 * GameType.java
 *
 * Created on October 12, 2003, 10:28 AM
 */

package com.jbolo.bolo;

/**
 * GameType is responsible for storing the game type (open, tournament strict
 * tournament) as well as calculating the number of supplies that are given to a
 * new tank
 * 
 * @author John Morrison
 */
public class GameType {

	/** An open game */
	public static final int OPEN_GAME_TYPE = 1;

	/** A tournament game */
	public static final int STRICT_GAME_TYPE = 3;

	/** A strict tournament game */
	public static final int TOURNAMENT_GAME_TYPE = 2;

	/** A full load of armour on the tank */
	private static int FULL_ARMOUR = 40;

	/** A full load of shells on the tank */
	private static int FULL_SHELLS = 40;

	/** A full load of mines on the tank */
	private static int FULL_MINES = 40;

	/** A full load of trees on the tank */
	private static int FULL_TREES = 40;

	/** The game type of this game */
	private int gameType;

	/** Game World this belongs to */
	private GameWorld gw;

	/**
	 * Creates a new instance of GameType
	 * 
	 * @param gw
	 *            GameWorld this belongs to
	 * @param gameType
	 *            The game type of this game
	 */
	public GameType(GameWorld gw, int gameType) {
		this.gameType = gameType;
		this.gw = gw;
	}

	/**
	 * Creates an instance of game type
	 * 
	 * @param gameType
	 *            The game type of the game
	 */
	void set(int gameType) {
		this.gameType = gameType;
	}

	/**
	 * getType returns the game type of the game
	 * 
	 * @return the game type of the game One of OPEN_GAME, STRICT_GAME,
	 *         TOURNAMENT_GAME
	 */
	public int getType() {
		return gameType;
	}

	/**
	 * Returns the number of shells a new tank gets
	 * 
	 * @return The amount of shells
	 */
	public int getShells() {
		int numBases;
		double percent;
		Bases bss;

		if (gameType == OPEN_GAME_TYPE) {
			return FULL_SHELLS;
		} else if (gameType == STRICT_GAME_TYPE) {
			return 0;
		} else {
			bss = gw.getBases();
			numBases = bss.getNumItems();
			if (numBases == 0) {
				numBases = 1;
			}
			percent = bss.getNumNeutral();
			percent /= numBases;
			return (int) (2.0 * percent);
		}
	}

	/**
	 * Returns the number of mines a new tank gets
	 * 
	 * @return The amount of mines
	 */
	public int getMines() {
		if (gameType == OPEN_GAME_TYPE) {
			return FULL_MINES;
		}
		return 0;
	}

	/**
	 * Returns the number of armour a new tank gets
	 * 
	 * @return The amount of armour
	 */
	public int getArmour() {
		return FULL_ARMOUR;
	}

	/**
	 * Returns the number of trees a new tank gets
	 * 
	 * @return The number of trees
	 */
	public int getTrees() {
		if (gameType == OPEN_GAME_TYPE) {
			return FULL_TREES;
		}
		return 0;
	}
}

File: winbolo/src/bolo/gametype.c
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */


/*********************************************************
*Name:          gametype
*Filename:      gametype.c
*Author:        John Morrison
*Creation Date: 29/1/99
*Last Modified:  6/1/00
*Purpose:
*  Responsable for tracking amounts of armour, health etc.
*  that should be given to a tank and gametype
*********************************************************/

#include <stdlib.h>
#include "global.h"
#include "screen.h"
#include "backend.h"
#include "gametype.h"

/* 2 as a define to eliminate magic number warnings */
#define TWO 2.0
/* Max bases = 16 as a double */
#define BASE_MAX_DOUBLE 16.0

gameType gmeType = gameOpen; /* Type of game being played */

/*********************************************************
*NAME:          gameTypeSet
*AUTHOR:        John Morrison
*CREATION DATE: 29/1/99
*LAST MODIFIED: 17/12/03
*PURPOSE:
* Sets the game type. Should be set before any other
* calls are made to the module
*
*ARGUMENTS:
*  gmeType      - Pointer to the game type
* value - Value to set the game type to
*********************************************************/
void gameTypeSet(gameType *gmeType, gameType value) {
  *gmeType = value;
}

/*********************************************************
*NAME:          gameTypeGet
*AUTHOR:        John Morrison
*CREATION DATE: 29/01/99
*LAST MODIFIED: 17/12/03
*PURPOSE:
* Returns the game type.
*
*ARGUMENTS:
*  gmeType      - Pointer to the game type
*********************************************************/
gameType gameTypeGet(gameType *gmeType) {
  return *gmeType;
}

/*********************************************************
*NAME:          gameTypeGetItems
*AUTHOR:        John Morrison
*CREATION DATE: 29/01/99
*LAST MODIFIED: 17/12/03
*PURPOSE:
* Called when a tank needs to restart. Fills the 
* parameters with how much stuff it should hold.
*
*ARGUMENTS:
*  gmeType      - Pointer to the game type
*  shellsAmount - Pointer to hold the number of shells
*  mines  - Pointer to hold the number of mines
*  armour - Pointer to hold the amount of armour
*  trees  - Pointer to hold the number of trees
*********************************************************/
void gameTypeGetItems(gameType *gmeType, BYTE *shellsAmount, BYTE *mines, BYTE *armour, BYTE *trees) {
  double percent; /* Percent of free bases */
  BYTE numBases;  /* Number of bases on the map */

  *armour = TANK_FULL_ARMOUR;
  switch (*gmeType) {
  case gameOpen:
    *shellsAmount = TANK_FULL_SHELLS;
    *mines = TANK_FULL_MINES;
    *trees = TANK_FULL_TREES;
    break;
  case gameTournament:
    numBases = screenNumBases();
    if (numBases == 0) {
      numBases = 1;
    }
    percent = (double) screenGetNumNeutralBases();
    percent /= numBases;
    percent *= BASE_MAX_DOUBLE;
    *shellsAmount = (BYTE) (TWO * percent);
    *mines = 0;
    *trees = 0;
    break;
  case gameStrictTournament:
  default:
    /* gameStrictTournament */
    *shellsAmount = 0;
    *mines = 0;
    *trees = 0;
    break;
  }
}
